app/services/samples/metadata/file_import_service.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+require 'roo'
+
+module Samples
+  module Metadata
+    # Service used to import sample metadata via a file
+    class FileImportService < BaseService
+      SampleMetadataFileImportError = Class.new(StandardError)
+
+      def initialize(project, user = nil, params = {})
+        super(user, params)
+        @project = project
+        @file = params[:file]
+        @sample_id_column = params[:sample_id_column]
+        @ignore_empty_values = params[:ignore_empty_values] || false
+        @spreadsheet = nil
+        @headers = nil
+      end
+
+      def execute
+        authorize! @project, to: :update_sample?
+
+        validate_sample_id_column
+
+        validate_file
+
+        perform_file_import
```

Comment:
On line 98, you're returning `response`, so I'm assuming you'd want to assign it to a variable here, and then presumably also return it to the controller (unless you're waiting to plug it into a controller first).

Reply:
My understanding is that the last line is always returned. The tests are able to evaluate the response. The controller should be able to call the service and receive the response. I think we're okay.